crates/red_knot_python_semantic/src/util/subscript.rs

```diff
@@ -1,83 +1,448 @@
-pub(crate) trait PythonSubscript {
+//! This module provides utility functions for indexing (`PyIndex`) and slicing
+//! operations (`PySlice`) on iterators, following the semantics of equivalent
+//! operations in Python.
+
+use std::cmp::Ordering;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(crate) struct OutOfBoundsError;
+
+pub(crate) trait PyIndex {
     type Item;
 
-    fn python_subscript(&mut self, index: i64) -> Option<Self::Item>;
+    fn py_index(&mut self, index: i32) -> Result<Self::Item, OutOfBoundsError>;
 }
 
-impl<I, T: DoubleEndedIterator<Item = I>> PythonSubscript for T {
-    type Item = I;
+enum Nth {
+    FromStart(usize),
+    FromEnd(usize),
+}
+
+fn from_nonnegative_i32(index: i32) -> usize {
+    static_assertions::const_assert!(usize::BITS >= 32);
+    debug_assert!(index >= 0);
 
-    fn python_subscript(&mut self, index: i64) -> Option<I> {
+    // SAFETY: `index` is non-negative, and `usize` is at least 32 bits.
+    usize::try_from(index).unwrap()
+}
+
+fn from_negative_i32(index: i32) -> usize {
+    static_assertions::const_assert!(usize::BITS >= 32);
+
+    index.checked_neg().map(from_nonnegative_i32).unwrap_or({
+        // 'checked_neg' only fails for i32::MIN. We can not
+        // represent -i32::MIN as a i32, but we can represent
+        // it as a usize, since usize is at least 32 bits.
+        from_nonnegative_i32(i32::MAX) + 1
+    })
+}
+
+impl Nth {
+    fn from_index(index: i32) -> Self {
         if index >= 0 {
-            self.nth(usize::try_from(index).ok()?)
+            Nth::FromStart(from_nonnegative_i32(index))
         } else {
-            let nth_rev = usize::try_from(index.checked_neg()?).ok()?.checked_sub(1)?;
-            self.rev().nth(nth_rev)
+            Nth::FromEnd(from_negative_i32(index) - 1)
+        }
+    }
+
+    fn to_nonnegative_index(&self, len: usize) -> usize {
+        debug_assert!(len > 0);
+
+        match self {
+            Nth::FromStart(nth) => *nth,
+            Nth::FromEnd(nth_rev) => (len - 1).saturating_sub(*nth_rev),
+        }
+    }
+}
+
+impl<I, T> PyIndex for T
+where
+    T: DoubleEndedIterator<Item = I>,
+{
+    type Item = I;
+
+    fn py_index(&mut self, index: i32) -> Result<I, OutOfBoundsError> {
+        match Nth::from_index(index) {
+            Nth::FromStart(nth) => self.nth(nth).ok_or(OutOfBoundsError),
+            Nth::FromEnd(nth_rev) => self.rev().nth(nth_rev).ok_or(OutOfBoundsError),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(crate) struct StepSizeZeroError;
+
+pub(crate) trait PySlice {
+    type Item;
+
+    fn py_slice<'a>(
+        &'a mut self,
+        start: Option<i32>,
+        stop: Option<i32>,
+        step: Option<i32>,
+    ) -> Result<Box<dyn Iterator<Item = Self::Item> + 'a>, StepSizeZeroError>
+    where
+        Self::Item: 'a;
+}
+
+impl<I, T> PySlice for T
```

Comment:
I haven't looked into the usages of the `PySlice` type but would it make sense to implement it for `&[T]` instead of implementing it on iterators? Or are there cases where we only have an iterator?

Reply:
Yes, thanks. We only use it on slices if we are okay with `collect`ing the chars in the `LiteralString` case.